src/Microsoft.AspNet.OData.Shared/Query/AutoSelectExpandHelper.cs

```diff
@@ -347,5 +334,95 @@ private static bool CanExpand(this IEdmModel edmModel, IEdmStructuredType struct
                 }
             }
         }
+
+        /// <summary>
+        /// This is a helper that allows us to avoid inefficiencies in the previous pattern:
+        ///     var structuredTypes = new List&lt;IEdmStructuredType&gt;();
+        ///     structuredTypes.Add(structuredType);
+        ///     structuredTypes.AddRange(edmModel.FindAllDerivedTypes(structuredType));
+        ///
+        /// Specifically, the allocation of the list and the resizing driven by the "AddRange" call are
+        /// avoided by leveraging a struct with a simple state machine for enumerating over a type
+        /// and its derived types.
+        /// </summary>
+        private struct SelfAndDerivedEnumerator : IEnumerator<IEdmStructuredType>
+        {
+            private readonly IEnumerator<IEdmStructuredType> derivedEnumerator;
+            private readonly IEdmStructuredType structuredType;
+
+            private byte stage;
+
+            public SelfAndDerivedEnumerator(IEdmStructuredType structuredType, IEdmModel edmModel)
+            {
+                if (structuredType == null)
+                {
+                    throw new ArgumentNullException(nameof(structuredType));
+                }
+
+                if (edmModel == null)
+                {
+                    throw new ArgumentNullException(nameof(edmModel));
+                }
+
+                this.stage = 0;
+                this.derivedEnumerator = edmModel.FindAllDerivedTypes(structuredType).GetEnumerator();
+                this.structuredType = structuredType;
+            }
+
+            public IEdmStructuredType Current
+            {
+                get
+                {
+                    switch (stage)
+                    {
+                        case 1:
+                            return this.structuredType;
+
+                        case 2:
+                            return this.derivedEnumerator.Current;
+
+                        default:
+                            throw new InvalidOperationException("Enumeration is an invalid state");
+                    }
+                }
+            }
+
+            object IEnumerator.Current => this.Current;
+
+            public void Dispose()
```

Comment:
I actually predict that this will perform better than `Prepend` at least because we are not disposing `this.derivedEnumerator`. Please add that logic in here.